Removed MyXactAccessedTempRel

xinzweb opened the pull request:
In Postgres, PREPARE TRANSACTION errors out if it accessed a temp relation.

In GPDB, two phase commit is always used, even for transactions accessing temp
relations, which is used by MPP query processing.

In this case, GPDB treats temp table as regular distributed table, but shorten
its life cycle to sessions or transactions.

GPDB also stores temp table information in shared buffers, hence MyXactAccessedTempRel
is not used, and we are removing it.

Signed-off-by: Xin Zhang <xzhang@pivotal.io>

asimrp commented on src/backend/access/transam/xact.c:
```diff
@@ -3713,27 +3714,20 @@ PrepareTransaction(void)
+	 * GPDB treat temporary table like a regular table, e.g. stored in shared buffer
+	 * instead of keep it in local buffer. The temporary table just have a shorter life
+	 * cycle either tie to the session or tie to the transaction if ON COMMIT clause is
+	 * used.
```
In addition, unlike PostgreSQL, transactions in GPDB should not remain in prepared state for arbitrarily long time.  PostgreSQL must depend on an external transaction coordinator to commit a prepared transaction.  In case of GPDB, QD backend acts as 2PC coordinator and sends COMMIT broadcast soon after PREPARE.  This mitigates the risk involved in letting a transaction block in proc_exit() for removing temp tables.

ashwinstar replied:
Good point Asim will include it too.

hlinnaka replied:
You have addressed the local buffer issue, but what about all the other problems mentioned in the Postgres comment?

xinzweb replied:
@hlinnaka I think the comments addressed both issues:
- Access of local buffers
- Backend blocked

Is there any other issues we haven't addressed? Thanks.

ashwinstar commented:
Lets remove setting / remaining reference in xact.c of MyXactAccessedTempRel=false

hlinnaka commented on src/backend/access/transam/xact.c:
```diff
@@ -81,11 +81,12 @@ int			CommitDelay = 0;	/* precommit delay in microseconds */
- * MyXactAccessedTempRel is set when a temporary relation is accessed.
- * We don't allow PREPARE TRANSACTION in that case.  (This is global
- * so that it can be set from heapam.c.)
+ * In Postgres, MyXactAccessedTempRel is used to prevent
```
I think I'd prefer to keep the variable, and the code in heapam.c to set it, even if we don't need it in GPDB, to minimize the diff vs. upstream and avoid merge conflicts.

xinzweb replied:
@hlinnaka How about we wrap the code with something like `#ifdef POSTGRES_MERGE`, where GPDB doesn't need it, but we keep the code to avoid merge conflict? So that, when people look at the code later, they understand the purpose of the code is NOT for functionality in GPDB, but just for merging.

hlinnaka commented:
On 01/18/2017 08:58 PM, Xin Zhang wrote:
>> @@ -81,11 +81,12 @@ int			CommitDelay = 0;	/* precommit delay in microseconds */
>  int			CommitSiblings = 5; /* # concurrent xacts needed to sleep */
>
>  /*
> - * MyXactAccessedTempRel is set when a temporary relation is accessed.
> - * We don't allow PREPARE TRANSACTION in that case.  (This is global
> - * so that it can be set from heapam.c.)
> + * In Postgres, MyXactAccessedTempRel is used to prevent
>
> @hlinnaka How about we wrap the code with something like `#ifdef
> POSTGRES_MERGE`, where GPDB doesn't need it, but we keep the code to
> avoid merge conflict? So that, when people look at the code later,
> they understand the purpose of the code is NOT for functionality in
> GPDB, but just for merging.

Note that if you put #ifdefs around the variable definition, you need to 
#ifdef out all the references to it, too. They will cause merge 
conflicts, just like removing the code. IMO it would nevertheless be 
slightly better than removing the code completely, as it would make the 
intention more clear (i.e. that the code is not needed in GPDB, rather 
than that it has been removed by accident). That makes the merge 
conflicts quicker to resolve, but they will still happen and need manual 
fixing.

Or you could leave the variable in place, perhaps with a comment to 
explain that the variable is not used for anything in GPDB.

- Heikki

xinzweb commented:
We'd like to come up a clear approach on marking unused code in GPDB for the case to help future Postgres Merge. 

We don't want to leave dead code behind, which is harder to maintain down the road, and also confused the intention of the functionality of the code. 

It's painful to deal with those merge conflicts, but considering the benefits of cleaner intention of the code, we'd like to proceed on marking the code out with `#ifdef POSTGRES_MERGE`.

danielgustafsson commented on src/backend/access/transam/xact.c:
```diff
@@ -3713,6 +3717,24 @@ PrepareTransaction(void)
+	 * used.
+	 *
+	 * Every transaction in GPDB is 2PC, so PREPARE TRANSACTION is used even for temp table
+	 * creation. GPDB cannot error out, other wise, it won't be able to handle temp table
```
s/other wise/otherwise/

danielgustafsson commented:
This might be bikeshed territory but I'd be inclined to use ifdef zero like below for code which doesn't apply but is left to make merges easier:
```C
#if 0 /* Upstream code not applicable to GPBD */
  ..
#endif
```
For anyone with syntax highlighting this will mark all the code as commented out making it easier to visually spot dead code when grepping around in a file. Including a nice descriptive comment as to why something is killed, as it's done in this PR, is of course a big plus and should be mandatory unless perhaps it's blindingly obvious.

xinzweb commented:
@danielgustafsson great idea. Updated the code. Thanks.

hlinnaka commented:
FWIW, you misspelled "GPDB" in the comments...

On 20 January 2017 20:28:32 EET, Xin Zhang <notifications@github.com> wrote:
>@danielgustafsson great idea. Updated the code. Thanks.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/1573#issuecomment-274143942

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

xinzweb commented:
Thanks for spotting it. Fixed.

On Fri, Jan 20, 2017 at 11:18 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> FWIW, you misspelled "GPDB" in the comments...
>
> On 20 January 2017 20:28:32 EET, Xin Zhang <notifications@github.com>
> wrote:
> >@danielgustafsson great idea. Updated the code. Thanks.
> >
> >--
> >You are receiving this because you were mentioned.
> >Reply to this email directly or view it on GitHub:
> >https://github.com/greenplum-db/gpdb/pull/1573#issuecomment-274143942
>
> --
> Sent from my Android device with K-9 Mail. Please excuse my brevity.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/1573#issuecomment-274155845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAsLujiUIp-29NeWuVj0bqBBI1vVZ8mCks5rUQh8gaJpZM4LmTeq>
> .
>



-- 
SHIN
